### PR TITLE
SUPESC-287 Fixed Klarna Invoice payment method.

### PR DESCRIPTION
**Pull request by boykodev**
- Developer(s): @boykodev

- Ticket: https://spryker.atlassian.net/browse/SUPESC-287

- Release Group: https://release.spryker.com/release-groups/view/3323

- Release Overview: https://release.spryker.com/release/pull-request-eco/Adyen/33

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Adyen                 | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Introduced `AdyenConstants::KLARNA_RETURN_URL` to configure return URL for Klarna payment method.
- Introduced `AdyenConfig::getKlarnaPayReturnUrl()`.
- Introduced `AdyenConfig::getKlarnaTaxRateMultiplier()`.
- Introduced `AdyenApiRequestConfig::REDIRECT_RESULT_FIELD`.
- Introduced `CallbackController::redirectKlarnaAction()` to handle redirect from Adyen.
- Adjusted `AdyenRouteProviderPlugin` to add redirect Klarna route.
- Adjusted `OnlineTransferRedirectHandler` to add redirect results to payment details.
- Adjusted `KlarnaInvoiceMapper` so it uses `AdyenConfig::getKlarnaTaxRateMultiplier()` config for tax multiplier.

### CODE REVIEW

File: src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php [hunk 1]
```diff
@@ -139,6 +139,7 @@ protected function getRequestDetails(AdyenRedirectResponseTransfer $redirectResp
     {
         return [
             AdyenApiRequestConfig::PAYLOAD_FIELD => $redirectResponseTransfer->getPayload(),
+            AdyenConfig::REDIRECT_RESULT_FIELD => $redirectResponseTransfer->getRedirectResult(),
```
Comment (asmarovydlo) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php:142]: Why it's not used directly from `\SprykerEco\Shared\Adyen\AdyenApiRequestConfig::REDIRECT_RESULT_FIELD`?

File: src/SprykerEco/Zed/Adyen/AdyenConfig.php [hunk 2]
```diff
@@ -274,6 +279,16 @@ public function getWeChatPayReturnUrl(): string
         return $this->get(AdyenConstants::WE_CHAT_PAY_RETURN_URL);
     }
 
+    /**
+     * @api
```
Comment (asmarovydlo) [inline src/SprykerEco/Zed/Adyen/AdyenConfig.php:283]: Please add specification.

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
Please add labels to the PR.
Please check changelogs:
1. Each line should have `.` at the end.
2. `Enabled Klarna payment method by default.` - what this change is about? Please get rid of it.
3. `.php` extension should not be present in the changelogs.
4.  Please add release overview link into the PR description. (See RG errors)
5. src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php this should be mentioned in the changelogs in the `Fixes` section. (Please find corresponding public API and describe changes through it).

**Comment by asmarovydlo** [comment]
Please follow the next pattern for PR names https://spryker.slack.com/archives/C03KVH4N8/p1597669214097700

**Comment by asmarovydlo** [comment]
Please move any touched eco module to Github Actions
See an example at https://github.com/spryker/event-behavior

File: src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php (hunk 1 above)
Comment (sadkivskyi) [reply src/SprykerEco/Zed/Adyen/Business/Handler/Redirect/OnlineTransferRedirectHandler.php]: Changed to `AdyenApiRequestConfig`

File: src/SprykerEco/Zed/Adyen/AdyenConfig.php (hunk 2 above)
Comment (sadkivskyi) [reply src/SprykerEco/Zed/Adyen/AdyenConfig.php:294]: Added

**Review by asmarovydlo — APPROVED** [review]

File: src/SprykerEco/Zed/Adyen/AdyenConfig.php [hunk 3]
```diff
@@ -287,6 +287,19 @@ public function getWeChatPayReturnUrl(): string
         return $this->get(AdyenConstants::WE_CHAT_PAY_RETURN_URL);
     }
 
+    /**
+     * Specification:
+     * - Returns the klarna payment method return url.
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Adyen/AdyenConfig.php:292]: Klarna's

File: src/SprykerEco/Shared/Adyen/AdyenConstants.php [hunk 4]
```diff
@@ -36,6 +36,14 @@ interface AdyenConstants
     public const IDEAL_ISSUERS_LIST = 'ADYEN:IDEAL_ISSUERS_LIST';
     public const REQUEST_CHANNEL = 'ADYEN:REQUEST_CHANNEL';
 
+    /**
+     * Specification:
+     * - Return URL for Klarna payment method.
+     *
+     * @api
+     */
+    public const KLARNA_RETURN_URL = 'ADYEN:KLARNA_RETURN_URL';
```
Comment (gechetspr) [inline src/SprykerEco/Shared/Adyen/AdyenConstants.php:45]: Are we sure that this value env specific?

File: src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php [hunk 5]
```diff
@@ -17,13 +17,17 @@
 class KlarnaInvoiceMapper extends AbstractMapper
 {
     protected const REQUEST_TYPE = 'klarna';
+    protected const KLARNA_TAX_RATE_MULTIPLIER = 100;
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php:20]: Looks like smth that we can put into a config

File: src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php [hunk 6]
```diff
@@ -17,13 +17,17 @@
...
 
     /**
      * @return string
      */
     protected function getReturnUrl(): string
     {
-        return '';
+        /** @var \SprykerEco\Zed\Adyen\AdyenConfig $config */
+        $config = $this->config;
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php:28]: We might can do this by `@method` annotation in this class

**Review by gechetspr — CHANGES_REQUESTED** [review]
Also please check changelogs

E.g. `Adjusted KlarnaInvoiceMapper so it uses const KLARNA_TAX_RATE_MULTIPLIER instead for int value.` is not completely true. We are using this constant to multiply previous value

File: src/SprykerEco/Shared/Adyen/AdyenConstants.php (hunk 4 above)
Comment (sadkivskyi) [reply src/SprykerEco/Shared/Adyen/AdyenConstants.php:45]: Yes like other return redirects.

File: src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php (hunk 5 above)
Comment (sadkivskyi) [reply src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php]: Moved to config

File: src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php (hunk 6 above)
Comment (sadkivskyi) [reply src/SprykerEco/Zed/Adyen/Business/Hook/Mapper/MakePayment/KlarnaInvoiceMapper.php]: Added property annotation

**Review by gechetspr — APPROVED** [review]

**Pull request merged**